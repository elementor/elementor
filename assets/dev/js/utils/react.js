import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { createRoot } from 'react-dom/client';

/**
 * Support conditional rendering of a React App to the DOM, based on the React version.
 * We use `createRoot` when available, but fallback to `ReactDOM.render` for older versions.
 *
 * @param { React.ReactElement } app        The app to render.
 * @param { HTMLElement }        domElement The DOM element to render the app into.
 *
 * @return {{ unmount: () => void }} The unmount function.
 */
function render( app, domElement ) {
	let unmountFunction;

	try {
		const root = createRoot( domElement );
		root.render( app );

		unmountFunction = () => {
			root.unmount();
		};
	} catch ( e ) {
		// eslint-disable-next-line react/no-deprecated
		ReactDOM.render( app, domElement );

		unmountFunction = () => {
			// eslint-disable-next-line react/no-deprecated
			ReactDOM.unmountComponentAtNode( domElement );
		};
	}

	return { unmount: unmountFunction };
}

export default {
	render,
};
