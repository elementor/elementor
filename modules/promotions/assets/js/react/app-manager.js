import App from './app';
import { createRoot } from 'react-dom/client';

export class AppManager {
	constructor() {
		this.promotionInfoTip = null;
		this.onRoute = () => {};
	}

	getPromotionData( promotionType ) {
		return elementorPromotionsData[ promotionType ] || {};
	}

	mount( targetNode, selectors ) {
		if ( this.promotionInfoTip ) {
			return;
		}

		const wrapperElement = targetNode?.closest( selectors.wrapperElement );
		const rootElement = wrapperElement?.querySelector( selectors.reactAnchor );

		if ( ! rootElement ) {
			return;
		}

		this.attachEditorEventListeners();

		this.promotionInfoTip = createRoot( rootElement );

		const colorScheme = elementor?.getPreferences?.( 'ui_theme' ) || 'auto';
		const isRTL = elementorCommon.config.isRTL;
		const promotionType = rootElement.getAttribute( 'data-promotion' )?.replace( '_promotion', '' );

		this.promotionInfoTip.render(
			<App
				colorScheme={ colorScheme }
				isRTL={ isRTL }
				promotionsData={ this.getPromotionData( promotionType ) }
				onClose={ () => this.unmount() }
			/>,
		);
	}

	unmount() {
		if ( this.promotionInfoTip ) {
			this.detachEditorEventListeners();
			this.promotionInfoTip.unmount();
		}

		this.promotionInfoTip = null;
	}

	attachEditorEventListeners() {
		this.onRoute = ( component, route ) => {
			if ( route !== 'panel/elements/categories' && route !== 'panel/editor/content' ) {
				return;
			}
			this.unmount();
		};

		$e.routes.on( 'run:after', this.onRoute );
	}

	detachEditorEventListeners() {
		$e.routes.off( 'run:after', this.onRoute );
	}
}
