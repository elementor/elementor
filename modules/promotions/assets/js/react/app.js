import {
	DirectionProvider,
	Infotip,
	LocalizationProvider,
	ThemeProvider,
} from '@elementor/ui';
import PromotionCard from './components/promotion-card';

const App = ( props ) => {
	return (
		<DirectionProvider rtl={ props.isRTL }>
			<LocalizationProvider>
				<ThemeProvider colorScheme={ props.colorScheme }>
					<Infotip
						content={ <PromotionCard doClose={ props.onClose } promotionsData={ props.promotionsData } /> }
						placement="right"
						arrow={ true }
						open={ true }
						disableHoverListener={ true }
						PopperProps={ {
							modifiers: [
								{
									name: 'offset',
									options:
										{ offset: [ -24, 8 ] },
								},
							],
						} }
					/>
				</ThemeProvider>
			</LocalizationProvider>
		</DirectionProvider>
	);
};

App.propTypes = {
	colorScheme: PropTypes.oneOf( [ 'auto', 'light', 'dark' ] ),
	isRTL: PropTypes.bool,
	promotionsData: PropTypes.object,
	onClose: PropTypes.func.isRequired,
};

export default App;
