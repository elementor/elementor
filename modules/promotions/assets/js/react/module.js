import PromotionControl from './controls/promotion';

export default class Module extends elementorModules.editor.utils.Module {
	onElementorInit() {
		elementor.addControlView( 'promotion_control', PromotionControl );
	}
}
