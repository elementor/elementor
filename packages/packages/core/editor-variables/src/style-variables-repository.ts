import { createStyleVariablesRepository } from './create-style-variables-repository';

export const styleVariablesRepository = createStyleVariablesRepository();
