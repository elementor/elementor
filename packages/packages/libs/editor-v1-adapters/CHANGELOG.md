# Change Log

## 0.12.1

### Patch Changes

- Updated dependencies [e953081]
  - @elementor/utils@0.5.0

## 0.12.0

### Minor Changes

- 7439d73: extract `isExperimentActive` to `@elementor/editor-v1-adapters`

## 0.11.0

### Minor Changes

- 4c2935b: Added support to atomic elements for command "document/elements/reset-style"

### Patch Changes

- e798985: Support preview in Global Classes

## 0.10.2

### Patch Changes

- 959e02c: Disable canvas and app bar buttons when opening the class manager panel

## 0.10.1

### Patch Changes

- Updated dependencies [571ff75]
  - @elementor/utils@0.4.0

## 0.10.0

### Minor Changes

- b8e2a85: Add data hooks support

### Patch Changes

- da38fa9: Remove unused exports and add missing dependencies.
- ad6fde0: Support history actions in style changes
- Updated dependencies [9ca4eab]
  - @elementor/utils@0.3.1

## 0.9.1

### Patch Changes

- a8b60c9: Make undoable synchronous and pass the updated redo result to undo

## 0.9.0

### Minor Changes

- 65d3c4c: Support sync & internal commands

## 0.8.5

### Patch Changes

- 91453b3: Update and lock dependencies versions

## 0.8.4

### Patch Changes

- 0c6bcb6: Introduce `undoable` function which allows creating an action with undo/redo support based on V1.
- Updated dependencies [e7f4706]
  - @elementor/utils@0.3.0

## 0.8.3

### Patch Changes

- 9abdfaf: Fix package.json exports field

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.8.2](https://github.com/elementor/elementor-packages/compare/@elementor/editor-v1-adapters@0.8.1...@elementor/editor-v1-adapters@0.8.2) (2024-08-20)

**Note:** Version bump only for package @elementor/editor-v1-adapters

## [0.8.1](https://github.com/elementor/elementor-packages/compare/@elementor/editor-v1-adapters@0.8.0...@elementor/editor-v1-adapters@0.8.1) (2024-08-05)

### Bug Fixes

- publish only necessary files to npm ([#226](https://github.com/elementor/elementor-packages/issues/226)) ([d808e2f](https://github.com/elementor/elementor-packages/commit/d808e2f60eb7ca2d7b8560d0b79c0e62c2f969a8))

# [0.8.0](https://github.com/elementor/elementor-packages/compare/@elementor/editor-v1-adapters@0.7.0...@elementor/editor-v1-adapters@0.8.0) (2024-07-16)

### Features

- **editor-editing-panel:** api for selecting and widget type [EDS-244] ([#199](https://github.com/elementor/elementor-packages/issues/199)) ([3e3da8a](https://github.com/elementor/elementor-packages/commit/3e3da8ac1dc37e502991f0905cbe8054422b3b71))

# [0.7.0](https://github.com/elementor/elementor-packages/compare/@elementor/editor-v1-adapters@0.6.2...@elementor/editor-v1-adapters@0.7.0) (2024-07-08)

### Features

- **editor-editing-panel:** create editor panel package [EDS-226] ([#188](https://github.com/elementor/elementor-packages/issues/188)) ([e361aed](https://github.com/elementor/elementor-packages/commit/e361aed023c3a5d4dd329b354f1403de238da20e))

## [0.6.2](https://github.com/elementor/elementor-packages/compare/@elementor/editor-v1-adapters@0.6.1...@elementor/editor-v1-adapters@0.6.2) (2024-07-03)

**Note:** Version bump only for package @elementor/editor-v1-adapters

## [0.6.1](https://github.com/elementor/elementor-packages/compare/@elementor/editor-v1-adapters@0.6.0...@elementor/editor-v1-adapters@0.6.1) (2024-07-02)

**Note:** Version bump only for package @elementor/editor-v1-adapters

# [0.6.0](https://github.com/elementor/elementor-packages/compare/@elementor/editor-v1-adapters@0.5.0...@elementor/editor-v1-adapters@0.6.0) (2023-09-14)

### Features

- **v1-adapters:** mark private functions [ED-12029] ([#122](https://github.com/elementor/elementor-packages/issues/122)) ([2e79ccc](https://github.com/elementor/elementor-packages/commit/2e79ccc87add5f0508b5a142f9f6d832b4657062))

# [0.5.0](https://github.com/elementor/elementor-packages/compare/@elementor/editor-v1-adapters@0.4.0...@elementor/editor-v1-adapters@0.5.0) (2023-06-11)

### Features

- **editor-panels:** init [ED-10804] ([#44](https://github.com/elementor/elementor-packages/issues/44)) ([1ed4113](https://github.com/elementor/elementor-packages/commit/1ed41131db8fb9151163175bfa614f784159e04b))

# [0.4.0](https://github.com/elementor/elementor-packages/compare/@elementor/editor-v1-adapters@0.3.0...@elementor/editor-v1-adapters@0.4.0) (2023-06-06)

### Features

- drop support for React 17 [ED-10982] ([#50](https://github.com/elementor/elementor-packages/issues/50)) ([59c576c](https://github.com/elementor/elementor-packages/commit/59c576ca218947dc0992616311d4d399a20e91a6))

# 0.3.0 (2023-06-01)

### Features

- add support for React 18 [ED-10905] ([#46](https://github.com/elementor/elementor-packages/issues/46)) ([c8915c6](https://github.com/elementor/elementor-packages/commit/c8915c6ea62550bcdeb5c8a576f311bedc35bcad))

# 0.2.0 (2023-05-09)
