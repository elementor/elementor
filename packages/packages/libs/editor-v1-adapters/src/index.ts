export {
	runCommand as __privateRunCommand,
	runCommandSync as __privateRunCommandSync,
	openRoute as __privateOpenRoute,
	registerRoute as __privateRegisterRoute,
} from './dispatchers/dispatchers';

export {
	useIsRouteActive as __privateUseIsRouteActive,
	useListenTo as __privateUseListenTo,
	useRouteStatus as __privateUseRouteStatus,
	type UseRouteStatusOptions,
} from './hooks';

export {
	setReady as __privateSetReady,
	listenTo as __privateListenTo,
	flushListeners as __privateFlushListeners,
	dispatchReadyEvent as __privateDispatchReadyEvent,
	windowEvent,
	v1ReadyEvent,
	commandStartEvent,
	commandEndEvent,
	routeOpenEvent,
	routeCloseEvent,
} from './listeners';

export type * from './listeners';

export { isRouteActive as __privateIsRouteActive, isExperimentActive } from './readers';

export { undoable } from './undoable';

export type { HistoryItem, WindowWithHistoryManager } from './undoable/get-history-manager';

export { useEditMode, changeEditMode, type EditMode } from './edit-mode';

export { registerDataHook } from './data-hooks/register-data-hook';
export { blockCommand } from './data-hooks/block-command';
