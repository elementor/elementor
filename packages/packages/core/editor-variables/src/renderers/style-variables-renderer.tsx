import * as React from 'react';
import { useEffect, useState } from 'react';
import { __privateUseListenTo as useListenTo, commandEndEvent } from '@elementor/editor-v1-adapters';
import { Portal } from '@elementor/ui';

import { styleVariablesRepository } from '../style-variables-repository';
import { getCanvasIframeDocument } from '../sync/get-canvas-iframe-document';
import { type StyleVariables, type Variable } from '../types';

const VARIABLES_WRAPPER = 'body';

export function StyleVariablesRenderer() {
	const container = usePortalContainer();
	const styleVariables = useStyleVariables();

	const hasVariables = Object.keys( styleVariables ).length > 0;

	if ( ! container || ! hasVariables ) {
		return null;
	}

	const cssVariables = convertToCssVariables( styleVariables );
	const wrappedCss = `${ VARIABLES_WRAPPER }{${ cssVariables }}`;

	return (
		<Portal container={ container }>
			<style data-e-style-id="e-variables" key={ wrappedCss }>
				{ wrappedCss }
			</style>
		</Portal>
	);
}

function usePortalContainer() {
	return useListenTo( commandEndEvent( 'editor/documents/attach-preview' ), () => getCanvasIframeDocument()?.head );
}

function useStyleVariables() {
	const [ variables, setVariables ] = useState< StyleVariables >( {} );

	useEffect( () => {
		const unsubscribe = styleVariablesRepository.subscribe( setVariables );

		return () => {
			unsubscribe();
		};
	}, [] );

	return variables;
}

function cssVariableDeclaration( key: string, variable: Variable ) {
	const variableName = variable?.deleted ? key : variable.label;
	const value = variable.value;

	return `--${ variableName }:${ value };`;
}

function convertToCssVariables( variables: StyleVariables ): string {
	const listOfVariables = Object.entries( variables );
	return listOfVariables.map( ( [ key, variable ] ) => cssVariableDeclaration( key, variable ) ).join( '' );
}
