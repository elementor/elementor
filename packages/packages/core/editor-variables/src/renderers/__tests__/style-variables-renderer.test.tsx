import * as React from 'react';
import { __privateUseListenTo as useListenTo } from '@elementor/editor-v1-adapters';
import { render, screen } from '@testing-library/react';

import { styleVariablesRepository } from '../../style-variables-repository';
import { StyleVariablesRenderer } from '../style-variables-renderer';

// Mock dependencies
const unsubscribeMock = jest.fn();

jest.mock( '@elementor/editor-v1-adapters', () => ( {
	__privateUseListenTo: jest.fn(),
	commandEndEvent: jest.fn(),
} ) );

jest.mock( '@elementor/ui', () => ( {
	Portal: jest.fn( ( { children } ) => <div data-testid="portal">{ children }</div> ),
} ) );

jest.mock( '../../style-variables-repository', () => ( {
	styleVariablesRepository: {
		subscribe: jest.fn( () => unsubscribeMock ),
	},
} ) );

jest.mock( '../../sync/get-canvas-iframe-document', () => ( {
	getCanvasIframeDocument: jest.fn(),
} ) );

describe( '<StyleVariablesRenderer />', () => {
	it( 'should not render anything when container is not available', () => {
		// Arrange.
		jest.mocked( useListenTo ).mockReturnValue( null );

		// Act.
		render( <StyleVariablesRenderer /> );

		// Assert.
		// eslint-disable-next-line testing-library/no-test-id-queries
		expect( screen.queryByTestId( 'portal' ) ).not.toBeInTheDocument();
	} );

	it( 'should render style variables in portal when container is available', () => {
		// Arrange.
		const mockContainer = document.createElement( 'div' );
		const mockVariables = {
			'e-gv-a1': { value: '#ff0000', label: 'red-color', type: 'global-color-variable' },
			'e-gv-a2': { value: 'Arial', label: 'font-arial', type: 'global-font-variable' },
			'e-gv-a3': { value: 'yellow', label: 'yellow-color', type: 'global-color-variable' },
			'e-gv-a4': { value: 'red', label: 'red-color', type: 'global-color-variable', deleted: true },
		};

		jest.mocked( useListenTo ).mockReturnValue( mockContainer );
		jest.mocked( styleVariablesRepository.subscribe ).mockImplementation( ( callback ) => {
			callback( mockVariables );

			return jest.fn();
		} );

		// Act.
		render( <StyleVariablesRenderer /> );

		// Assert.
		// eslint-disable-next-line testing-library/no-test-id-queries
		expect( screen.getByTestId( 'portal' ) ).toMatchSnapshot();
	} );

	it( 'should not render anything when there are no style variables', () => {
		// Arrange.
		const mockContainer = document.createElement( 'div' );

		jest.mocked( useListenTo ).mockReturnValue( mockContainer );
		jest.mocked( styleVariablesRepository.subscribe ).mockImplementation( ( callback ) => {
			callback( {} );

			return jest.fn();
		} );

		// Act.
		render( <StyleVariablesRenderer /> );

		// Assert.
		// eslint-disable-next-line testing-library/no-test-id-queries
		expect( screen.queryByTestId( 'portal' ) ).not.toBeInTheDocument();
	} );

	it( 'should unsubscribe from style variables on unmount', () => {
		// Arrange.
		const mockContainer = document.createElement( 'div' );

		jest.mocked( useListenTo ).mockReturnValue( mockContainer );
		jest.mocked( styleVariablesRepository.subscribe ).mockImplementation( () => unsubscribeMock );

		// Act.
		const { unmount } = render( <StyleVariablesRenderer /> );

		unmount();

		// Assert.
		expect( unsubscribeMock ).toHaveBeenCalled();
	} );
} );
