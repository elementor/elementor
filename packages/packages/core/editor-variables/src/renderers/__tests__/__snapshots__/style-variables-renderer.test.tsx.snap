// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StyleVariablesRenderer /> should render style variables in portal when container is available 1`] = `
<div
  data-testid="portal"
>
  <style
    data-e-style-id="e-variables"
  >
    body{--red-color:#ff0000;--font-arial:Arial;--yellow-color:yellow;--e-gv-a4:red;}
  </style>
</div>
`;
